TU20Bot/Commands/NameMatch.cs

```diff
@@ -0,0 +1,242 @@
+
+                        // Person exists in the server
+                        addIfNotContains(indexNotToRead, index);
+
+                        return;
+                    }
+
+                    // For testing purposes
+                    discordMessage.Append($"`{firstName} {lastName}` full name matched.\n");
+
+                    // Person exists in the server
+                    addIfNotContains(indexNotToRead, index);
+                },
+
+                /** If the last name matches but the first name is different **/
+                (index, firstName, lastName) =>
+                {
+                    // If first name doesn't match
+                    errorLog.Append($"`{firstName} {lastName}`'s last name matches " +
+                            $"with `{listNames[index, firstNameIndex]} {listNames[index, lastNameIndex]}`'s last name.\n");
+
+                    // Person exists with some name in the server
+                    addIfNotContains(indexNotToRead, index);
+                });
+            }
+
+            // For adding all the messages to send to the user
+            discordMessage.Append(errorLog);
+
+            // For creating a list of all the indexes not present in the indexNotToRead List
+            for (int i = 0; i < listNames.GetLength(0); i++)
+            {
+
+                // For creating a list that has indexes to read from for the not present in server message
+                if (!indexNotToRead.Contains(i))
+                {
+                    // For reporting if the user doesn't exist in the server (no last name match)
+                    discordMessage.Append($"{listNames[i, firstNameIndex]} {listNames[i, lastNameIndex]}" +
+                                          $" did not have a last name match therefore is not in the server.\n");
+                }
+
+            }
+
+            return discordMessage.ToString();
+        }
+
+        private static void addIfNotContains<T>(List<T> list, T value)
+        {
+            if (!list.Contains(value)) list.Add(value);
+        }
+
+        /// <summary>
+        /// Given an unknown name and a set of valid names, identify any matches.
+        /// </summary>
+        /// <param name="fullName">An unknown name</param>
+        /// <param name="listNames">A set peoples names dived into first and last name</param>
+        /// <param name="firstOrLastMatch">Executed when there is no space in fullname, however either the first or last name match</param>
+        /// <param name="succesfulMatch">Executed when there is a complete match, that is both first name and last name match</param>
+        /// <param name="lastNameMatch">Executed when only the last name matches </param>
+        public static void nameMatchAlg(string fullName, string[,] listNames, IndexMatch firstOrLastMatch, IndexMatchDetailed succesfulMatch, IndexMatchDetailed lastNameMatch)
```

Comment:
Sorry @viktorKorolyuk this delegate system might need a bit more explaining to me. They seem like they are working around a return value, which would generally give more control over to the parent function and potentially increase the usefulness of the algorithm.

I suggest:
```
enum NameMatchLevel {
  // NoMatch ?
  NoSpacesMatch,
  LastNameMatch,
  CompleteMatch
}

class NameMatchProfile {
  NameMatchLevel level;
  string fullName;
}

public static List<NameMatchProfile> matchNames(string fullName, ListName[] names) { ... }
```

The additional control can come in the form of queries to the list. For example:
```
var matches = matchNames(...);

matches.IsEmpty() // no matches

matches.FirstOrDefault(x => x.level == NameMatchLevel.CompleteMatch) // find the first complete match
matches.Count(x => x.level == NameMatchLevel.CompleteMatch) // how many people fully matched?

matches.Count(x => x.name.StartsWith("John")) // analyze names
```

Generally, this can also be done with delegates, but its more complicated. Variables need to be used to track specfic information. And it seems like the tests pass empty lambdas just so this function can be called.

Reply:
I like that idea! If we create a list from the parsed data the file structure will be flatter and the related logic can be grouped together even more.